src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+
+            bool isAssumedValidEnabled = this.consensusSettings.BlockAssumedValid != null;
+            bool isWithinCheckpoints = this.consensusSettings.UseCheckpoints && (earliestNewHeader.Height <= this.checkpoints.GetLastCheckpointHeight()); ;
+
+            if (isWithinCheckpoints || isAssumedValidEnabled)
+            {
+                ChainedHeader currentChainedHeader = latestNewHeader;
+
+                while (currentChainedHeader != earliestNewHeader)
+                {
+                    if (currentChainedHeader.HashBlock == this.consensusSettings.BlockAssumedValid)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' represents an assumed valid block.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleAssumedValidHeader(currentChainedHeader, latestNewHeader);
+                        break;
+                    }
+
+                    CheckpointInfo checkpoint = this.checkpoints.GetCheckpoint(currentChainedHeader.Height);
+                    if (checkpoint != null)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' is a checkpoint.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleCheckpointsHeader(currentChainedHeader, latestNewHeader, checkpoint);
+                        break;
+                    }
+
+                    currentChainedHeader = currentChainedHeader.Previous;
+                }
+
+                if ((connectedHeaders == null) && isWithinCheckpoints)
+                {
+                    connectedHeaders = new ConnectedHeaders() {Consumed = latestNewHeader};
+                    this.logger.LogTrace("Chained header '{0}' bellow last checkpoint.", currentChainedHeader);
+                }
+
+                if (connectedHeaders != null)
+                {
+                    this.logger.LogTrace("(-)[CHECKPOINT_OR_ASSUMED_VALID]:{0}", connectedHeaders);
+                    return connectedHeaders;
+                }
+            }
+
+            if (latestNewHeader.ChainWork > this.chainState.ConsensusTip.ChainWork)
+            {
+                this.logger.LogDebug("Chained header '{0}' is the tip of a chain with more work than our current consensus tip.", latestNewHeader);
+
+                connectedHeaders = this.MarkBetterChainAsRequired(latestNewHeader);
+            }
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// A chain with more work than our current consensus tip was found so mark all it's descendants as required.
+        /// </summary>
+        /// <param name="latestNewHeader">The new header that represents a longer chain.</param>
+        /// <returns>The new headers that need to be downloaded.</returns>
+        private ConnectedHeaders MarkBetterChainAsRequired(ChainedHeader latestNewHeader)
```

Comment:
maybe `MarkBetterChainAsRequired` => `MarkAncestorAndSelfAsRequired`

Because MarkBetterChain kind of assumes that we will check the chainwork inside this method, but we dont

Reply:
Kind of agree, maybe I would go with just Mark chain is required. Also that argument should be renamed to something like chain tip.

Reply:
I like `MarkChainAsRequired`, that sounds better